httpcore5-h2/src/test/java/org/apache/hc/core5/http2/examples/Http2GreetingServer.java

```diff
@@ -0,0 +1,199 @@
+                                                                                            entityDetails,
+                                                                                    final HttpContext context) {
+            // if there's no body don't try to parse entity:
+            AsyncEntityConsumer entityConsumer = new NoopEntityConsumer();
+
+            if (entityDetails != null) {
+                entityConsumer = new StringAsyncEntityConsumer();
+            }
+            return new BasicRequestConsumer<String>(entityConsumer);
+
+        }
+
+        @Override
+        protected void handle(final Message<HttpRequest, String> requestMessage,
+                              final AsyncServerRequestHandler.ResponseTrigger responseTrigger,
+                              final HttpContext context) throws HttpException, IOException {
+
+            try {
+                final HttpCoreContext coreContext = HttpCoreContext.adapt(context);
+                final EndpointDetails endpoint = coreContext.getEndpointDetails();
+                final HttpRequest req = requestMessage.getHead();
+                final String httpEntity = requestMessage.getBody();
+
+                // generic success response:
+                final HttpResponse resp = new BasicHttpResponse(200);
+
+                // recording the request
+                System.out.println(String.format("[%s] %s %s %s", new Date(),
+                        endpoint.getRemoteAddress().toString(),
+                        req.getMethod(),
+                        req.getPath()));
+
+                // Request without an entity - GET/HEAD/DELETE
+                if (httpEntity == null) {
+                    responseTrigger.submitResponse(new BasicResponseProducer(resp), context);
+                    return;
+                }
+
+                // Request with an entity - POST/PUT
+                final ContentType entityContentType = ContentType.parse(req.getHeader(HttpHeaders.CONTENT_TYPE).getValue());
+                String greeting = "Hello stranger\n";
+
+                if (entityContentType.getMimeType().equals(ContentType.APPLICATION_FORM_URLENCODED.getMimeType())) {
+
+                    // decoding the form entity into key/value pairs:
+                    final List<NameValuePair> args = URLEncodedUtils.parse(httpEntity,
+                            ContentType.APPLICATION_FORM_URLENCODED.getCharset());
+
+                    // composing greeting:
+                    greeting = String.format("Hello %s\n", args.get(0).getValue());
+                }
+
+                responseTrigger.submitResponse(new BasicResponseProducer(resp,
+                        new StringAsyncEntityProducer(greeting, ContentType.TEXT_PLAIN)), context);
+            } catch (final HttpException | UnsupportedCharsetException | IOException e) {
+                throw new IllegalStateException(e);
```

Comment:
@rferreira Much better, but I still dislike exception handling code. There is no need to re-throw  `HttpException` and  `IOException` as `IllegalStateException`. Those exceptions do not represent illegal service state. They are perfectly legal. Either handle them in your handler and return a valid HTTP message back to the caller, or let the framework do it for you.